[FrameworkBundle][HttpClient] Add the retry_failed.base_uris option to http_client

Tiriel opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        |
| License       | MIT

`RetryableHttpClient` has been able to spread its attempts over several base URIs since HttpClient 6.3, through `shiftBaseUri()`. No FrameworkBundle configuration ever reached it: the scoped client `base_uri` node is a scalar, and `default_options` has no base URI node at all.

```yaml
framework:
    http_client:
        scoped_clients:
            my_api.client:
                base_uri: 'https://a.example.com'
                headers: { X-Api-Key: '%env(API_KEY)%' }
                retry_failed:
                    base_uris:
                        - 'https://b.example.com'
```

`base_uris` lists the URIs to fall back to after `base_uri`, one per attempt. It lives under `retry_failed` because the rotation only advances between attempts: it is a property of the retry policy, not of the client.

Two things stand in the way of simply passing the list down, and both are handled here.

**The decorator order.** `ScopingHttpClient::request()` resolves the request URL against the scope base URI and then does `unset($options['base_uri'])`. Anything below it receives an absolute URL and no base URI, so a base URI injected by the retryable client is never applied and every attempt goes to the same host. The retryable client is now decorated at priority 12 instead of 25 **when and only when `base_uris` is set**. Every other configuration keeps the container it has today, byte for byte; I dumped and diffed a container built from the existing `http_client_retry` fixture to confirm it.

One consequence: at priority 12 the retryable client also sits outside `CachingHttpClient`, so retries re-enter the cache. Failures are not cached and a rotated URI has a different cache key, so a retry misses and reaches the transport as before, and a cache hit no longer pays for the retry machinery.

**The scope.** The regexp that `ScopingHttpClient::forBaseUri()` derives from a single base URI stops matching as soon as the host rotates, so the scoped options, the API key above for instance, would silently stop applying on the fallback host. A new `ScopingHttpClient::forBaseUris()` derives a regexp covering all of them. It has to live in the component: the derivation is `resolveUrl(parseUrl('.'), parseUrl($baseUri))`, which drops the last path segment, and those helpers are private to `HttpClientTrait`. A naive `preg_quote` gets it wrong for any base URI with a path:

```
https://a.example.com/v1/api   derived: https\://a\.example\.com/v1/
                                naive: https\://a\.example\.com/v1/api/
```

Also fixed: the default client now receives the list too. It was only wired at the scoped call site, so `framework.http_client.default_options.retry_failed.base_uris` validated, compiled, and did nothing.

Configurations that would silently do nothing are rejected rather than accepted: `base_uris` without a `base_uri`, and `base_uris` with retries disabled.

Tested through a real request rather than on the service definitions. `HttpClientTest::testScopedClientRetriesAgainstTheNextBaseUri` records what reaches the transport and asserts both the rotation and that the scoped header survives it:

```
GET https://a.example.com/posts/1 X-Scoped: yes
GET https://b.example.com/posts/1 X-Scoped: yes
```

With the retryable client left at its old priority the second line reads `a.example.com`, and with `forBaseUri()` in place of `forBaseUris()` the scoped header disappears from it. Neither is visible to a definition-level assertion.

Tiriel commented:
Not sure about the xml conf, advices welcome!

nicolas-grekas commented:
Can you please rebase on 6.3 since this is a bugfix?

Tiriel committed:
[FrameworkBundle] Fix config for array of base_uri in http_client

Allow array of `base_uri` in config for HttpClient keys, as per
[comment in PR](https://github.com/symfony/symfony/pull/49809#issuecomment-1860729153).

Tiriel committed:
Fix indent

Tiriel commented:
Rebased on 6.3!

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1901,13 +1901,17 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode, callable $e
                                 ->end()
+                                ->validate()
+                                    ->ifTrue(fn ($v) => (isset($v['base_uri']) && \is_array($v['base_uri'])) && !isset($v['retry_failed']))
+                                    ->thenInvalid('"base_uri" can only be an array if "retry_failed" is defined.')
```
defined or enabled for retry_failed?
also, should we validate the array contains only strings?

Tiriel replied:
I'd say that more precautions are always better, so indeed checking if there are strings only!
And yeah, retries should be enabled, imo.

Going to make changes in that direction.

javiereguiluz replied:
In this code we're merging all validations but the error message is the same for all. If you define an array and one of the values is no a string by mistake, you'll see this and will be confusing:

```
"base_uri" can only be an array if "retry_failed" is defined
```

Maybe we could extract this validation and create a dedicated error message for it?

```php
\count($v['base_uri']) !== \count(array_filter($v['base_uri'], 'is_string'))
```

Thanks!

Tiriel committed:
Base uri argument for `RetryableHttpClient::__construct`

Add optional `$baseUri` argument for `RetryableHttpClient::__construct`
to allow proper integration in franework

Tiriel commented:
Made an attempt just as a discussion base. Works as is, but maybe there's a better way to do it or some constraints I haven't tought about.

Tiriel committed:
Fix scopeConfig array for base_uri

Tiriel committed:
Fix CS from patch

Tiriel committed:
Rework to avoid refactoring of RetryableHttpClient

Tiriel commented:
Okay there have been obviously a huge mistake on my part here... Fixing asap.

Tiriel commented:
Seems okay with the right destination branch. Sorry for the spam!

Tiriel commented:
Test failures seem unrelated here, but any advice would be welcome!

Tiriel commented:
Hey there!

Bumping this, as it's still a missing feature. Should I create a new PR to get back in the recent ones?

nicolas-grekas commented:
Hi @Tiriel, thanks for your patience.
I think I changed my mind on the topic: this looks like a new feature to me.
To built on this idea, I'm wondering if it wouldn't make more sense to allow a new `base_uris` option under `retry_failed`.
That would make everything simpler I think.

nicolas-grekas commented on src/Symfony/Component/HttpClient/RetryableHttpClient.php:
```diff
@@ -39,12 +39,13 @@ class RetryableHttpClient implements HttpClientInterface, ResetInterface
      * @param int $maxRetries The maximum number of times to retry
      */
-    public function __construct(HttpClientInterface $client, ?RetryStrategyInterface $strategy = null, int $maxRetries = 3, ?LoggerInterface $logger = null)
+    public function __construct(HttpClientInterface $client, ?RetryStrategyInterface $strategy = null, int $maxRetries = 3, ?LoggerInterface $logger = null, array $baseUris = [])
```
this change is not needed nor used

Tiriel commented:
Also going to rebase and target the latest branch asap

nicolas-grekas commented:
I think base_uris under retry_strategy is more flexible as it allows using scope on top to control how scoped clients behave

Tiriel commented:
Then how do we deal with the `base_uri` option under each scoped client? Unless I miss something, it would be possible to define a `base_uri` in a scope client and an array of `base_uris` in the `retry_strategy`.

I see four possibilities:
* adding a `base_uri` under the scoped client and an array of `base_uris` in the retry strategy is blocked, raising an exception
* adding a `base_uri` under the scoped client and an array of `base_uris` in the retry strategy is *not* blocked, base uri from scoped client is kept
* adding a `base_uri` under the scoped client and an array of `base_uris` in the retry strategy is *not* blocked, base uris from retry strategy are kept
* adding a `base_uri` under the scoped client and an array of `base_uris` in the retry strategy is *not* blocked, base uri from scoped client is merged inside the array

Personnally, I think the first option (blocking) or the last (merging) are the best, with a preference for the blocking.

nicolas-grekas commented:
I would merge them: the base_uri is prepended to base_uris if any.

nicolas-grekas commented:
Hum, actually no: we'd need to make setting both base_uri and base_uris conflict, and we'd then need to compute a regular expression that "`|`" all base_uris so that scoped clients accepts them all in its scope (unless `scope` is set)

nicolas-grekas commented:
Alternatively, if we keep base_uri as allowing an array, we could make defining a `retry_strategy` optional (and auto-enable it then).
We also need to allow setting both a `scope` and an array of `base_uris` for completeness (both options are exclusive at the moment).

WDYT?

Tiriel commented:
> Alternatively, if we keep base_uri as allowing an array, we could make defining a `retry_strategy` optional (and auto-enable it then).

This seems like a better option to me, a bit more friendly to use

> We also need to allow setting both a `scope` and an array of `base_uris` for completeness (both options are exclusive at the moment).

This seems a bit redundant. IIRC, `scope` makes the configuration active only on uris matching the scope. So if you set an array of base uris and the scope, the scope will simply be active on every call since the base uris matching the scope will be prepended to each call.

In this case, I haven't looked at feasability yet, but maybe we could have a unified `scope`/`base_uri` option, working differently depending on the actual call you make: when calling `request`, if you pass a full uri with scheme/host, it works as scope, and if you pass a path, it prepends the base uri.

Seems a lot of work though, and not necessarily on the scope of this PR.

Tiriel committed:
[HttpClient] Add base_uris option under retry_failed for scoped clients

Allow scoped HTTP clients to define multiple base URIs for retry rotation
via retry_failed.base_uris. When more than one URI is provided, an explicit
scope must be defined on the scoped client to cover all retry URIs.

The base_uris option is inherited from default_options.retry_failed and can
be overridden per scoped client, following the same propagation rules as
other retry_failed options.

mtarld commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/yml/http_client_retry.yml:
```diff
@@ -16,3 +16,9 @@ framework:
                 base_uri: http://example.com
                 retry_failed:
                     multiplier: 4
+            bar:
```
Maybe you missed `http_client_retry.xml`?

Tiriel replied:
This is entirely possible :sweat_smile:

Tiriel replied:
Although, hasn't xml config been removed in 8.1? (new target branch)

mtarld replied:
Indeed, you're right!

mtarld commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -2044,13 +2044,23 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode, callable $e
                                     ->end()
-                                    ->scalarNode('base_uri')
-                                        ->info('The URI to resolve relative URLs, following rules in RFC 3985, section 2.')
+                                    ->variableNode('base_uri')
```
```php
->variableNode('base_uri')
    ->beforeNormalization()
        ->ifTrue(fn ($v) => null !== $v && !\is_string($v) && !(\is_array($v) && \count($v) !== \count(array_filter($v, 'is_string')))
        ->thenInvalid('"base_uri" must be a string or an array of strings.')
    ->end()
```

mtarld commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -2044,13 +2044,23 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode, callable $e
                                     ->ifTrue(static fn ($v) => !empty($v['query']) && !isset($v['base_uri']))
                                     ->thenInvalid('"query" applies to "base_uri" but no base URI is defined.')
                                 ->end()
+                                ->validate()
```
What about:
```php
->validate()
    ->ifTrue(fn ($v) => isset($v['base_uri']) && \is_array($v['base_uri']) && (!isset($v['retry_failed']) || false === $v['retry_failed']['enabled']))
    ->thenInvalid('"base_uri" can only be an array if "retry_failed" is enabled.')
->end()
->validate()
    ->ifTrue(fn ($v) => isset($v['base_uri']) && \is_array($v['base_uri']) && \count($v['base_uri']) !== \count(array_filter($v['base_uri'], 'is_string')))
    ->thenInvalid('"base_uri" array must only contain strings.')
->end()
```
to make things clearer?

Tiriel replied:
I left this because I don t think it's relevant anymore since `base_uri` can't be an array anymore in the configuration (cf constructor argument comment)

mtarld commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -2935,12 +2935,14 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
             if (null === $scope) {
-                $baseUri = $scopeConfig['base_uri'];
-                unset($scopeConfig['base_uri']);
+                $baseUri = \is_array($scopeConfig['base_uri']) ? $scopeConfig['base_uri'][0] : $scopeConfig['base_uri'];
```
When base_uri is an array, only the first entry is used by `ScopingHttpClient` to match requests.

What about requiring a `scope` option when `base_uri` is an array, so users explicitly define the matching regex instead of silently falling back to the first URI?

Something like:
```php
->validate()
    ->ifTrue(fn ($v) => isset($v['base_uri']) && \is_array($v['base_uri']) && !isset($v['scope']))
    ->thenInvalid('"scope" is required when "base_uri" is an array.')
->end()
```

Tiriel replied:
That's a fair point, I've tried something along this way last week but not committed yet. Going to push it to get your thoughts

Tiriel replied:
Implemented

mtarld commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -3048,7 +3055,8 @@ private function registerRetryableHttpClient(array $options, string $name, Conta
             ->setArguments([new Reference('.inner'), $retryStrategy, $options['max_retries'], new Reference('logger')])
-            ->addTag('monolog.logger', ['channel' => 'http_client']);
+            ->addTag('monolog.logger', ['channel' => 'http_client'])
+            ->addMethodCall('withOptions', [['base_uri' => $baseUris]], true);
```
I wonder if the `$baseUris` constructor parameter can be removed.

IIUC, `withOptions(['base_uri' => $baseUris])` already sets `$this->baseUris`, and `->addMethodCall(?, ?, true)` replaces the instance, therefore having both paths is redundant to me.

Tiriel replied:
Removed here, IMO it was indeed redundant and outdated

Tiriel committed:
Address reviewer feedback: scope required for any base_uris, remove stale is_array guard

Require an explicit scope whenever retry_failed.base_uris is set (not
only for >1 entries), so users always define the regex that covers all
retry URIs instead of silently relying on base_uri fallback.

Remove the is_array guard on $scopeConfig['base_uri'] in FrameworkExtension
which was a leftover from the old design where base_uri itself could be
an array.

Add a ConfigurationTest asserting the new validation message.

Tiriel committed:
Fix CS: use static fn in array_filter closure

Tiriel committed:
[FrameworkBundle][HttpClient] Add the retry_failed.base_uris option to http_client

RetryableHttpClient can spread its attempts over several base URIs, but no
configuration reached that feature. The option lists the URIs to fall back to
after the base one, one per attempt.

Two things stand in the way of simply passing the list down. ScopingHttpClient
resolves the request URL against the scope base URI and then unsets the
"base_uri" option, so a base URI injected by a retryable client sitting below it
is never applied and every attempt goes to the same host; the retryable client
is therefore decorated further out when, and only when, base_uris is set, which
leaves the container of every other configuration untouched. And the scope
regexp that forBaseUri() derives from a single URI stops matching once the host
rotates, so the scoped options would silently stop applying; forBaseUris()
derives a regexp covering all of them instead.
